src/java/com/puppetlabs/puppetserver/pool/JRubyPool.java

```diff
@@ -1,234 +1,371 @@
+        lock.lock();
+        try {
+            final Thread currentThread = Thread.currentThread();
+            while (!isWriteLockHeldByCurrentThread(currentThread)) {
+                if (!isWriteLockHeld()) {
+                    writeLockThread = currentThread;
+                } else {
+                    notWriteLocked.await();
+                }
+            }
+            try {
+                while (registeredElements.size() != liveQueue.size()) {
+                    allRegisteredInQueue.await();
+                }
+            } catch (Exception e) {
+                freeWriteLock();
+                throw e;
+            }
+        } finally {
+            lock.unlock();
+        }
     }
 
-    /**
-    * Releases the write lock.
-    */
     @Override
-    public void unlock() throws InterruptedException {
-        lock.writeLock().unlock();
+    public void unlock() {
+        final ReentrantLock lock = this.lock;
+        lock.lock();
+        try {
+            final Thread currentThread = Thread.currentThread();
+            if (!isWriteLockHeldByCurrentThread(currentThread)) {
+                String lockErrorMessage;
+                if (isWriteLockHeldByAnotherThread(currentThread)) {
+                    lockErrorMessage = "held by " + writeLockThread;
+                } else {
+                    lockErrorMessage = "not held by any thread";
+                }
+                throw new IllegalStateException(
+                        "Unlock requested from thread not holding the lock.  " +
+                        "Requested from " +
+                        currentThread +
+                        " but lock " +
+                        lockErrorMessage +
+                        ".");
+            }
+            freeWriteLock();
+        } finally {
+            lock.unlock();
+        }
     }
 
-    /**
-     * @return a set of all of the known elements that have been registered with
-     *         this pool.
-     */
     public Set<E> getRegisteredElements() {
         return registeredElements;
     }
 
+    private void addFirst(E e) {
+        liveQueue.addFirst(e);
+        signalNotEmpty();
+    }
+
+    private void freeWriteLock() {
+        writeLockThread = null;
+        // Need to use 'signalAll' here because there might be multiple
+        // waiters (e.g., multiple borrowers) queued up, waiting for the
+        // pool to be unlocked.
+        notWriteLocked.signalAll();
+        // Borrowers that are woken up when an instance is returned to the
+        // pool and the write lock is held would then start waiting on a
+        // 'notWriteLocked' signal instead.  Re-signalling 'notEmpty' here
+        // allows any borrowers still waiting on the 'notEmpty' signal to be
+        // reawoken when the write lock is released, compensating for any
+        // 'notEmpty' signals that might have been essentially ignored from
+        // when the write lock was held.
+        if (liveQueue.size() > 0) {
```

Comment:
No, I wouldn't see any harm with that.  This just eliminates a little extra overhead in unnecessarily waking up borrowers that wouldn't be able to do anything but go right back to sleep in the case that the queue might be empty when the write lock is released.  I don't have a strong opinion on this one.  If @cprice404 feels strongly that this would be an improvement, I could eliminate the extra if check.